Enhanced Substitution of strings

LMS0815 opened the pull request:
Enhanced Substitution of strings for variable capabilities to enable
better 3rd party transcoding

Commit:
Enhanced Substitution of strings

Enhanced Substitution of strings for variable capabilities to enable
better 3rd party transcoding

Commit:
Some new player models and enhancements of old pic

Added some new player models and ehanced current images to meet the look
& feel of the modern GUI

mherger commented on convert.conf:
```diff
@@ -43,34 +43,38 @@
 # F - can transcode from a named file
 # R - can transcode from a remote URL (URL types unspecified)
 # 
-# O - can seek to a byte offset in the source stream (not yet implemented)
```
You removed the "not yet implemented" part, but didn't implement it, did you? Has this been implemented since the comment was added?

LMS0815 replied:
I copied/synchronized the comments between convert.conf and TranscodingHelper.pm
"not yet implemented" was not within the comments of TranscodingHelper.pm
I assume no one implemented those options. I reverted the changes for those lines and updated the commments in TranscodingHelper.pm too.

LMS0815 replied:
Hi Michael,

is there anything I can do to cat my patch merged soon?

Regards

Heiko

Commit:
ReAdded  (not yet implemented)

Commit:
Updated to current version of the files (32bit)

Commit:
Fixed BUG for stdin substitution

Changed stdin substitition from "-" to -
"-" is not accepded at moste codecs and creating an ERROR message.

Commit:
Revert "Fixed BUG for stdin substitution"

This reverts commit c59f15a8fa74a67e2552d488453b8fa252d929d4.

Commit:
BUG FIX for [stdin] substitution

Changed from "-" to - (Without Quotes)

Commit:
Substitution of ${filename}$ against contents

Extended substitution:

Substitution of ${filename}$ against the contents of {filename}

LMS0815 commented:
Final major update:

The filename to contents substitution, in conjunction with then new $channels$ ... substitions, will enable you to create "dynamic" transcoding like:

flc mp3 \* *
    # FB:{BITRATE=--compression %B.%Q1}D:{RESAMPLE=--rate %d}T:{TRIM=trim %t}U:{UNTIL= trim %t =%v}
    [sox] -q --guard --type flac $FILE$ --type mp3  $RESAMPLE$ $BITRATE$       - $TRIM$ $UNTIL$ ${sox_remix_$CHANNELS$_to_$OCHANNELS$.effects}$ ${sox_equalizer_$CLIENTID$.effects}$ ${sox_equalizer_$CLIENTID$.effects}$ 

where the remix and equilizer files are loaded depending on the variables.

${sox_remix_$CHANNELS$to_$OCHANNELS$.effects}$ 
-> ${sox_remix_1to_2.effects}$ 
-> remix -m  1v0.8409 1v0.8409

${sox_equalizer_$CLIENTID$.effects}$
-> ${sox_equalizer_00-0f-2e-00-9b-24.effects}$
-> loudness

${sox_equalizer_$CLIENTID$.effects}$ 
-> ${sox_equalizer_baby.effects}$ 
-> earwax

The new line for transcoding would be:
[sox] -q --guard --type flac $FILE$ --type mp3  $RESAMPLE$ $BITRATE$       - $TRIM$ $UNTIL$ remix -m  1v0.8409 1v0.8409 loudness earwax

mherger commented:
Thanks a lot for this patch! A few comments/questions:
- are the sox/flac binary updates required? I'm very reluctant to replace binaries unless there's a strong need for it.
- the icons have some ugly artifacts. If you want to have a clean PNG file to start with, grab it from https://github.com/Logitech/squeezeplay/blob/public/7.8/assets/MasterIcons/icon_blank.png
- the Jive logo used for the squeezeplay.png is owned by a different company (http://www.jivesoftware.com)
- would it be possible to separate the conversion changes from the others?

Thanks!

Commit:
Revert "Some new player models and enhancements of old pic"

This reverts commit b0bce9eb0942b26f4ed5ed994f79df18ec0a2286.

LMS0815 commented:
The sox update is essential.
The old version is buggy and crashes during seek (T:{TRIM=trim %t}U:{UNTIL= trim %t =%v}).
This was a long lasting bug which was addressed in the old convert.conf file, where lame/flac is chained with  sox  as a workaround. This is not needed any longer

Release Notes 14.4.1:
  ...
  o Fix crash when seeking within a FLAC file. [3476843](Eric Wong)

The flac update is essential to handle big multichannel files
Releasenote 1.3.0:
   ...
  o Read and write appropriate channel masks for 6.1 and 7.1 surround input WAV files. Documentation was also updated.

I have withdrawn the logo updates. They are not important.

Commit:
Updatet comments

added file substitution comment

Commit:
Commen harmonization

cosmetic updates

LMS0815 commented on Slim/Player/TranscodingHelper.pm:
```diff
@@ -516,26 +528,54 @@ sub tokenizeConvertCommand2 {
 		elsif ($v eq 'D') {$value = ($transcoder->{'samplerateLimit'} || 44100) / 1000;}
 		
-		elsif ($v eq 'f') {$value = '"' . $filepath . '"';}
+		elsif ($v eq 'f') {$value = ($filepath eq '-' ? $filepath : '"' . $filepath . '"');}
```
BUG FIX

LMS0815 commented on Slim/Player/TranscodingHelper.pm:
```diff
@@ -516,26 +528,54 @@ sub tokenizeConvertCommand2 {
-	$subs{'FILE'} = '"' . $filepath . '"';
-	$subs{'URL'} = '"' . $fullpath . '"';
-	$subs{'QUALITY'} = $quality;
+	$subs{'FILE'}     = ($filepath eq '-' ? $filepath : '"' . $filepath . '"');
```
BUG FIX

Commit:
added fractal notation for quality %Q substitution

to pass quality to the codec sox uses fractal notation:
-- compression 320.9 low quality
-- compression 320.0 DEFAULT
-- compression 320.01 best quality

Commit:
code cosmetics

tailing trimed

Commit:
ADD Substitution for OutputChannelCount

added Substitution for OutputChannelCount

Commit:
Optimization

Removed Qoutes for CLIENTID substitution.
Added File Create if substitition file was not found.

Commit:
File read optimization

Commit:
speed optimization

speed optimization be rearanging code and reusing calculation results

Commit:
replaced redundant calculation

LMS0815 commented:
The code was fully tested. No problems found within current installations.
All new features are tested as well using SOX to transcode (down-mix and DRC) with a several players and songs.
If you like to have a description of the down-mix and DRC implementation I am happy to share. 
Just advice how.

mherger commented:
@LMS0815 - the problem I have with your patch is that it requires new binaries. I'd have to build them for all supported platforms, not only Windows. Plus we've been using patched versions of those. We'd have to verify the new binaries don't break what those patches did fix...

Would there be a possibility to provide the binaries as part of your plugin, replacing the LMS own versions?

LMS0815 commented:
Using the patch as is will work with the old binaries as well.
You just would not be able to benefit from some of the new features.
To use this new features, you need do create your own private "custom-convert.conf" anyway.
Removing the binaries wont harm.

Commit:
Revert "Updated to current version of the files (32bit)"

This reverts commit 3766e3aac4b3b67a77d08efd84cb7ccc504601ab.

LMS0815 commented:
Binaries removed from this patch, Never the less I recommend - due to the bug-fixes - to plan a public update.

_I appreciate that you are taking care of quality_ :clap:

LMS0815 commented:
@mherger

the binaries are removed from the patch.
Anyone who is interested in deep diving on sound quality
Down/Up Mix and/or Digital Room Correction
should be able to download and install the latest codecs anyway.

Commit:
Improved logging

removed 'die'
improved log messages

Commit:
Added LOG flags

changed warn to info

Commit:
Withdrawn Code Cosmetics

LMS0815 commented:
Unfortunately I cannot withdraw the code cosmetics for TranscodingHelper.pm this is conflicting with current updates.

mherger commented:
I reverted those changes manually. Please git pull the latest and verify I didn't mess up anything. Thanks!

mherger commented:
@LMS0815 - Could you please provide some documentation for your enhancements? I'm tempted to revert your changes, as they add complexity for no good to the public. From today's point of view it seems to cover some of your rather exotic needs. But lack of documentation nobody else can take advantage of it.

Thanks for your understanding!

LMS0815 commented:
I you like to revert the changes, feel free to do so.

I carefully addend long missing information (and not complexity) to 
enable advanced Filter processing.
But based on my very last experience with contributing to LMS I expected 
the mail you just send since long.

The complex part, the Equalizer and Room correction add-on I never 
published because I was expecting the behavior you just shown.


Am 30.12.2016 um 04:00 schrieb mherger:
>
> @LMS0815 <https://github.com/LMS0815> - Could you please provide some 
> documentation for your enhancements? I'm tempted to revert your 
> changes, as they add complexity for no good to the public. From 
> today's point of view it seems to cover some of your rather exotic 
> needs. But lack of documentation nobody else can take advantage of it.
>
> Thanks for your understanding!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Logitech/slimserver/pull/22#issuecomment-269726060>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AH2ozqMpBx4TKMr4OISCOaiM_KqUm0bIks5rNHPUgaJpZM4CQwWi>.
>
